helpers/postgresql.go

```diff
@@ -0,0 +1,121 @@
+package helpers
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	_ "github.com/jackc/pgx/v5/stdlib" // PostgreSQL driver for ready check.
+)
+
+const (
+	postgresqlImage    = "docker.io/library/postgres:14.2"
+	postgresqlPort     = "5432"
+	postgresqlUser     = "eelbot"
+	postgresqlPassword = "testpassword"
+	postgresqlTimeout  = 10 * time.Second
+)
+
+// StartPostgreSQL starts and returns the URL to connect to a PostgreSQL instance. It first checks if the
+// "TEST_POSTGRESQL_URL" environment variable is set, and if it is, it just returns that (useful for connecting to a
+// locally running instance for debugging purposes). Otherwise, it spins up a new PostgreSQL instance via docker and
+// returns the URL to connect to it.
+func StartPostgreSQL() (uri string, close CloseFunc, err error) {
+	// Check the "TEST_POSTGRESQL_URL" environment variable first.
+	if u, ok := os.LookupEnv("TEST_POSTGRESQL_URL"); ok {
+		uri, close = u, func() {}
+		return
+	}
+
+	// Start a PostgreSQL instance via docker.
+	var cli *client.Client
+	if cli, err = client.NewClientWithOpts(client.FromEnv); err != nil {
+		return
+	}
+
+	ctx := context.Background()
+	var reader io.ReadCloser
+	if reader, err = cli.ImagePull(ctx, postgresqlImage, types.ImagePullOptions{}); err != nil {
```

Comment:
Curious, in the go appdev world, is this a typical thing to do.
Have the app bootstrap it's resources by programmatically bringing up containers for other apps it needs etc.. or is it something you are just experimenting with since its eelbot? 

Can't say I've seen this being done anywhere

Reply:
I've seen it been done before. It was either this or doing it in a script, and this seemed like the better way to go. How have you seen it done elsewhere?